man/lkbib.man
.\" -*- nroff -*-
.ds g \" empty
.ds G \" empty
.TH \*GLKBIB @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
\*glkbib \- search bibliographic databases
.SH SYNOPSIS
.B \*glkbib
[
.B \-v
]
[
.BI \-i fields
]
[
.BI \-p filename
]
[
.BI \-t n
]
.IR key \|.\|.\|.
.SH DESCRIPTION
.B \*glkbib
searches bibliographic databases for references that contain the keys
.IR key \|.\|.\|.
and prints any references found on the standard output.
.B \*glkbib
will search any databases given by
.B \-p
options, and then a default database.
The default database is taken from the
.SB REFER
environment variable if it is set,
otherwise it is
.BR @DEFAULT_INDEX@ .
For each database
.I filename
to be searched,
if an index
.IB filename @INDEX_SUFFIX@
created by
.BR \*gindxbib (@MAN1EXT@)
exists, then it will be searched instead;
each index can cover multiple databases.
.SH OPTIONS
.TP
.B \-v
Print the version number.
.TP
.BI \-p filename
Search
.IR filename .
Multiple
.B \-p
options can be used.
.TP
.BI \-i string
When searching files for which no index exists,
ignore the contents of fields whose names are in
.IR string .
.TP
.BI \-t n
Only require the first
.I n
characters of keys to be given.
Initially
.I n
is 6.
.SH ENVIRONMENT
.TP \w'\fBREFER'u+2n
.SB REFER
Default database.
.SH FILES
.TP \w'\fB@DEFAULT_INDEX@'u+2n
.B @DEFAULT_INDEX@
Default database to be used if the
.SB REFER
environment variable is not set.
.IB filename @INDEX_SUFFIX@
Index files.
.SH "SEE ALSO"
.BR \*grefer (@MAN1EXT@),
.BR \*glookbib (@MAN1EXT@),
.BR \*gindxbib (@MAN1EXT@)


NEWS
This file describes recent user-visible changes in groff.  Bug fixes
are not described.  There are more details in the man pages.

Groff
-----

o `-mFOO' will now search first for `FOO.tmac' and then for `tmac.FOO'.  The
  old behaviour has been changed to overcome problems with platforms which
  have an 8+3 file name limit, and platforms which have other versions of
  troff installed also. Additionally, all macro files have been renamed
  using the latter scheme to avoid 8+3 name clashes.

o The new environment variable GROFF_BIN_PATH will be checked for programs
  groff is calling (preprocessors, troff, and output devices) before PATH.
  If not set, it defaults to the directory where the groff binary is
  located.  Previously, it was PATH only.  The nroff script will only use
  GROFF_BIN_PATH to find the groff binary but passes both the GROFF_BIN_PATH
  and PATH environment variable to groff.

Troff
-----

o The mdoc package has been completely rewritten, using the full power of
  GNU troff to remove limitations of Unix troff (which is no longer
  supported).  Most important changes are:

  . No argument limit.
  . Almost all macros are parsed and callable (if it makes sense).
  . `.Lb': prints library names
  . `.Nm <punctuation>' now works as expected; `.Nm "" <punctuation>' has
    been withdrawn.
  . Updated `.St' command
  . `.Fx': prints FreeBSD
  . `.Ox': prints OpenBSD
  . `.Bsx': prints BSD/OS
  . `.Brq', `.Bro', `.Brc': brace enclosure macros
  . `.Bd -centered': center lines
  . `.Bl -xwidth <string>': interpret <string> and use the resulting width
  . support for double-sided printing (-rD1 command line switch)
  . support for 11pt and 12pt document sizes (-rS11, -rS12 command line
    switches)

  `groff_mdoc.man' replaces `groff_mdoc.samples.man'; it now completely
  documents the mdoc package.

  Great care has been taken to assure backwards compatibility.  If you
  encounter any abnormal results, please report them to bug-groff@gnu.org.

o A new command line option for the `man' macros (similar to the `mdoc'
  package has been implemented: `-rcR=1' (now the default in nroff mode) will
  produce one single, very long page instead of multiple pages.  `-rcR=0'
  will deactivate it.

o The `return' request has been added to return immediately from a macro.

o A new request `nop' (no operation) has been added which is similar to
  `if 1'.  For example,

    .if t \{\
    Hallo!
    .\}

  can now be written as

    .if t \{\
    .  nop Hallo!
    .\}

o `box' and `boxa' are two new requests which behave similarly to `di' and
  `da' but don't include a partially filled line (which is restored after
  ending the diversion).

o The `asciify' request has been extended to `unformat' space characters
  and some other escape sequences also.

  `\ ' will no longer be unformatted as a space but remains an unpaddable,
  unbreakable space character.

o The new `unformat' request is similar to `asciify' but only handles space
  characters and tabs specially if the diversion is reread, retaining font
  information.  This makes it possible to reformat diversions; for example
  the following

    .ll 3i
    .
    a01 a02 a03 a04 a05 a06 a07 a08 a09 a10.
    .
    .box box1
    .ev 1
    .nf
    \f[B]b01 b02 b03 b04 b05 b06 b07 b08 b09 b10.\f[P]
    .br
    .ev
    .box
    .
    c01 c02 c03 c04 c05 c06 c07 c08 c09 c10.
    .
    .unformat box1
    .box1

  gives

    a01  a02  a03  a04 a05 a06 a07
    a08 a09 a10.  c01 c02 c03  c04
    c05  c06 c07 c08 c09 c10.  b01
    b02 b03 b04 b05  b06  b07  b08
    b09 b10.

  Without the `unformat' request, space characters are converted to word
  space nodes which are no longer stretchable, and the result would be

    a01  a02  a03  a04 a05 a06 a07
    a08 a09 a10.  c01 c02 c03  c04
    c05  c06 c07 c08 c09 c10.  b01
    b02 b03 b04 b05 b06 b07 b08
    b09 b10.

o The new request `linetabs' controls the `line-tabs' mode.  In line-tabs
  mode, tab distances are computed relative to the (current) output line.
  Otherwise they are taken relative to the input line.  For example, the
  following

    .ds x a\t\c
    .ds y b\t\c
    .ds z c
    .ta 1i 3i
    \*x
    \*y
    \*z

  yields

    a         b         c

  In line-tabs mode, the same code gives

    a         b                   c

  The new read-only number register `.linetabs' returns 1 if in line-tabs
  mode, and 0 otherwise.

o Two new requests `tm1' and `tmc' have been added to improve writing
  messages to the terminal.  `tm1' is similar to `tm' but allows leading
  whitespace.  `tmc' is similar to `tm1' but doesn't emit a final newline.

o A new request `dei' (define indirect) has been added.  The first and
  second parameter of `dei' are taken from string registers rather than
  directly; this very special request is needed to make `trace.tmac'
  independent from the escape character (which might even be disabled).

o It is now possible to save and restore the escape character with two new
  requests `ecs' and `ecr'.

o The new escape sequence \B'...' is an analogon to `\A': If the string
  within the delimiters is a valid numeric expression, return character `1',
  and `0' otherwise.

o The new escape sequence `\:' inserts a zero-width break point.  This is
  similar to `\%' but without a soft hyphen character.

o The `tr' request can now map characters onto `\~'.

o Calling the `fam' request without an argument switches back to the
  previous font family.

o The new read-only register `.int' is set to a positive value if the last
  output line is interrupted (i.e., if it contains `\c').

o The `writem' request is not new, but hasn't been documented before.  This
  is similar to `write' but instead of a string the contents of a given
  macro or string is written to a stream.

o The read/write number register `hp' to get/set the current horizontal
  position relative to the input line isn't new but hasn't been documented
  properly before.


src/utils/lkbib/lkbib.cc
// -*- C++ -*-
/* Copyright (C) 1989-1992, 2000, 2001 Free Software Foundation, Inc.
     Written by James Clark (jjc@jclark.com)

This file is part of groff.

groff is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.

groff is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with groff; see the file COPYING.  If not, write to the Free Software
Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <assert.h>

#include "lib.h"
#include "errarg.h"
#include "error.h"

#include "defs.h"
#include "refid.h"
#include "search.h"

extern "C" const char *Version_string;

static void usage(FILE *stream)
{
  fprintf(stream, "usage: %s [-nv] [-p database] [-i XYZ] [-t N] keys ...\n",
	  program_name);
}

int main(int argc, char **argv)
{
  program_name = argv[0];
  static char stderr_buf[BUFSIZ];
  setbuf(stderr, stderr_buf);
  int search_default = 1;
  search_list list;
  int opt;
  static const struct option long_options[] = {
    { "help", no_argument, 0, CHAR_MAX + 1 },
    { "version", no_argument, 0, 'v' },
    { NULL, 0, 0, 0 }
  };
  while ((opt = getopt_long(argc, argv, "nvVi:t:p:", long_options, NULL))
	 != EOF)
    switch (opt) {
    case 'V':
      verify_flag = 1;
      break;
    case 'n':
      search_default = 0;
      break;
    case 'i':
      linear_ignore_fields = optarg;
      break;
    case 't':
      {
	char *ptr;
	long n = strtol(optarg, &ptr, 10);
	if (n == 0 && ptr == optarg) {
	  error("bad integer `%1' in `t' option", optarg);
	  break;
	}
	if (n < 1)
	  n = 1;
	linear_truncate_len = int(n);
	break;
      }
    case 'v':
      {
	printf("GNU lkbib (groff) version %s\n", Version_string);
	exit(0);
	break;
      }
    case 'p':
      list.add_file(optarg);
      break;
    case CHAR_MAX + 1: // --help
      usage(stdout);
      exit(0);
      break;
    case '?':
      usage(stderr);
      exit(1);
      break;
    default:
      assert(0);
    }
  if (optind >= argc) {
    usage(stderr);
    exit(1);
  }
  char *filename = getenv("REFER");
  if (filename)
    list.add_file(filename);
  else if (search_default)
    list.add_file(DEFAULT_INDEX, 1);
  if (list.nfiles() == 0)
    fatal("no databases");
  int total_len = 0;
  int i;
  for (i = optind; i < argc; i++)
    total_len += strlen(argv[i]);
  total_len += argc - optind - 1 + 1; // for spaces and '\0'
  char *buffer = new char[total_len];
  char *ptr = buffer;
  for (i = optind; i < argc; i++) {
    if (i > optind)
      *ptr++ = ' ';
    strcpy(ptr, argv[i]);
    ptr = strchr(ptr, '\0');
  }
  search_list_iterator iter(&list, buffer);
  const char *start;
  int len;
  int count;
  for (count = 0; iter.next(&start, &len); count++) {
    if (fwrite(start, 1, len, stdout) != (size_t)len)
      fatal("write error on stdout: %1", strerror(errno));
    // Can happen for last reference in file.
    if (start[len - 1] != '\n')
      putchar('\n');
    putchar('\n');
  }
  return !count;
}
